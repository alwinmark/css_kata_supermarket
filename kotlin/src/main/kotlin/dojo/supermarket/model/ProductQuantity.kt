package dojo.supermarket.model

class ProductQuantity(val product: Product, val quantity: Double)
