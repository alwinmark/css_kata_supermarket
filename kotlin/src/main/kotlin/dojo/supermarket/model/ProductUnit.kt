package dojo.supermarket.model

enum class ProductUnit {
    Kilo, Each
}
