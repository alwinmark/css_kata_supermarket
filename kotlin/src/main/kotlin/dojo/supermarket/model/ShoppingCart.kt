package dojo.supermarket.model

import java.util.*

class ShoppingCart {

    private var items: LinkedList<ShoppingCartItem> = LinkedList()
    private var productQuantities: MutableMap<Product, Double> = HashMap()


    internal fun getItems(): List<ShoppingCartItem> {
        return items
    }

    fun productQuantities(): Map<Product, Double> {
        return productQuantities
    }

    fun addShoppingCartItem(shoppingCartItem: ShoppingCartItem) {
        items.push(shoppingCartItem)
        val product = shoppingCartItem.product
        if (productQuantities.containsKey(product)) {
            productQuantities[product] = productQuantities[product]!! + shoppingCartItem.quantity
        } else {
            productQuantities[product] = shoppingCartItem.quantity
        }
    }
}
