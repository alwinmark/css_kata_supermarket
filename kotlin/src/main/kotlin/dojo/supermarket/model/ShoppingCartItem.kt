package dojo.supermarket.model

class ShoppingCartItem(val product: Product, val quantity: Double)