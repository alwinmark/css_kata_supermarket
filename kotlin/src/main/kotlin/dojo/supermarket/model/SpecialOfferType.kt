package dojo.supermarket.model

enum class SpecialOfferType {
    ThreeForTwo, TenPercentDiscount, TwoForAmount, FiveForAmount
}
