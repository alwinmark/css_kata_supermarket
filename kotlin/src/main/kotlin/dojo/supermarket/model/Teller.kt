package dojo.supermarket.model

import java.util.HashMap


class Teller(private val catalog: SupermarketCatalog) {
    private val offers = HashMap<Product, Offer>()

    fun addSpecialOffer(offerType: SpecialOfferType, product: Product, argument: Double) {
        this.offers[product] = Offer(offerType, product, argument)
    }

    fun checksOutArticlesFrom(cart: ShoppingCart): Receipt {
        val receipt = Receipt()
        val shoppingCartItems = cart.getItems()
        for (item in shoppingCartItems) {
            val p = item.product
            val quantity = item.quantity
            val unitPrice = catalog.getUnitPrice(item.product)
            val price = quantity * unitPrice
            receipt.addProduct(p, quantity, unitPrice, price)
        }
        handleOffers(receipt, cart)

        return receipt
    }

    private fun handleOffers(receipt: Receipt, cart: ShoppingCart) {
        val productQuantities = cart.productQuantities()
        for (item in productQuantities.keys) {
            val quantity = productQuantities[item]!!
            if (offers.containsKey(item)) {
                val offer = offers[item]!!
                val unitPrice = catalog.getUnitPrice(item)
                val quantityAsInt = quantity.toInt()
                var discount: Discount? = null
                var x = 1
                if (offer.offerType === SpecialOfferType.ThreeForTwo) {
                    x = 3

                } else if (offer.offerType === SpecialOfferType.TwoForAmount) {
                    x = 2
                    if (quantityAsInt >= 2) {
                        val total = offer.argument * (quantityAsInt / x) + quantityAsInt % 2 * unitPrice
                        val discountN = unitPrice * quantity - total
                        discount = Discount(item, "2 for " + offer.argument, discountN)
                    }

                }
                if (offer.offerType === SpecialOfferType.FiveForAmount) {
                    x = 5
                }
                val numberOfXs = quantityAsInt / x
                if (offer.offerType === SpecialOfferType.ThreeForTwo && quantityAsInt > 2) {
                    val discountAmount =
                        quantity * unitPrice - (numberOfXs.toDouble() * 2.0 * unitPrice + quantityAsInt % 3 * unitPrice)
                    discount = Discount(item, "3 for 2", discountAmount)
                }
                if (offer.offerType === SpecialOfferType.TenPercentDiscount) {
                    discount =
                        Discount(item, offer.argument.toString() + "% off", quantity * unitPrice * offer.argument / 100.0)
                }
                if (offer.offerType === SpecialOfferType.FiveForAmount && quantityAsInt >= 5) {
                    val discountTotal =
                        unitPrice * quantity - (offer.argument * numberOfXs + quantityAsInt % 5 * unitPrice)
                    discount = Discount(item, x.toString() + " for " + offer.argument, discountTotal)
                }
                if (discount != null)
                    receipt.addDiscount(discount)
            }

        }
    }
}
