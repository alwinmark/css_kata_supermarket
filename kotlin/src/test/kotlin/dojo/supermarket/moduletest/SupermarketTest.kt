package dojo.supermarket.moduletest

import dojo.supermarket.ReceiptPrinter
import dojo.supermarket.TestUtil.Companion.apples
import dojo.supermarket.TestUtil.Companion.toothbrush
import dojo.supermarket.model.*
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class SupermarketTest {


    @Test
    fun testSomething() {
        val catalog = FakeCatalog()
        catalog.addProduct(toothbrush, 0.99)
        catalog.addProduct(apples, 1.99)

        val cart =
            ShoppingCart()
        cart.addItemQuantity(apples, 2.5)

        val teller = Teller(catalog)
        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, toothbrush, 10.0)

        val receipt = teller.checksOutArticlesFrom(cart)

        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun testEmptyCatalogAndCart() {
        // arrange
        val emptyCatalog = FakeCatalog()
        val emptyCart = ShoppingCart()

        val teller = Teller(emptyCatalog)

        // act
        val receipt = teller.checksOutArticlesFrom(emptyCart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun testEmptyCart() {
        // arrange
        val catalog = FakeCatalog()
        catalog.addProduct(apples, 1.99)
        val emptyCart = ShoppingCart()

        val teller = Teller(catalog)

        // act
        val receipt = teller.checksOutArticlesFrom(emptyCart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }
    @Test
    fun addProductNotInCatalog() {
        // arrange
        val emptyCatalog  = FakeCatalog()
        val cart = ShoppingCart()
        cart.addItemQuantity(apples, 0.5)

        val teller = Teller(emptyCatalog)

        // act & assert
        assertThrows<NullPointerException> {
            teller.checksOutArticlesFrom(cart)
        }
    }

    @Test
    fun testProductWithoutDiscount() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 3.0)

        val teller = Teller(catalog)

        // act
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun testProductWithNegativeQuantity() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, -2.0)

        val teller = Teller(catalog)

        // act
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun testHalfToothbrush() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(toothbrush, 1.99)
        cart.addItemQuantity(toothbrush, 0.5)

        val teller = Teller(catalog)

        // act
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun testNegativePrice() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, -1.99)
        cart.addItemQuantity(apples, 2.0)

        val teller = Teller(catalog)

        // act
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun threeForTwo_onlyTwo() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 2.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun threeForTwo_three() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 3.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun threeForTwo_seven() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 7.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun addExistingProduct() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 7.0)
        val teller = Teller(catalog)

        // act
        cart.addItemQuantity(apples, 7.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun twoForAmount_One() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 1.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun twoForAmount_Two() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 2.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun twoForAmount_Seven() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 7.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun twoForAmount_NegativeAmount() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, -7.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun twoForNegativeAmount_Seven() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 7.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, apples, -2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun fiveForAmount_One() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 1.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun fiveForAmount_Five() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 5.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun fiveForAmount_Eleven() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 11.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun fiveForAmount_NegativeAmount() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, -7.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun fiveForNegativeAmount_Eleven() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 11.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, -2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun tenPercentDiscount_Eleven() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 11.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun tenPercentDiscount_NegativeAmount() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, -11.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, apples, 2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun tenPercentDiscount_Eleven_NegativeDiscount() {
        // arrange
        val catalog = FakeCatalog()
        val cart = ShoppingCart()
        catalog.addProduct(apples, 1.99)
        cart.addItemQuantity(apples, 11.0)

        val teller = Teller(catalog)

        // act
        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, apples, -2.0)
        val receipt = teller.checksOutArticlesFrom(cart)

        // assert
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }
}
