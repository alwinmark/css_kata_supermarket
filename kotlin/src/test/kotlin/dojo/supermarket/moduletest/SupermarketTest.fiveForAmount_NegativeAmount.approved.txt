apples                            -13.93
  1.99 * -7.000

Total:                            -13.93